drivers/vcluster/vcluster.go

```diff
@@ -103,3 +163,373 @@ func WaitForVClusterRunning(vclusterName string, timeout time.Duration) error {
+func (v *VCluster) CreateNodePortService() error {
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "vcluster-nodeport",
+			Namespace: v.Namespace,
+		},
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{
+				"app":     "vcluster",
+				"release": v.Name,
+			},
+			Ports: []v1.ServicePort{
+				{
+					Name:       "https",
+					Port:       443,
+					TargetPort: intstr.FromInt(8443),
+					Protocol:   v1.ProtocolTCP,
+				},
+			},
+			Type: v1.ServiceTypeNodePort,
+		},
+	}
+	createdSvc, err := k8sCore.CreateService(service)
+	if err != nil {
+		return err
+	}
+	v.NodePort = createdSvc.Spec.Ports[0].NodePort
+	return nil
+}
+
+// SetClientSetForVCluster method calculates the clientset for vcluster. This also
+// takes care of setting any extra params like skip-tls-verify, etc for vcluster
+func (v *VCluster) SetClientSetForVCluster() error {
+	serverURL := fmt.Sprintf("--server=https://%s:%d", ControlNodeIP, v.NodePort)
+	cmd := exec.Command("vcluster", "connect", v.Name, "-n", v.Namespace, "--update-current=false", serverURL)
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	currentDir, _ := os.Getwd()
+	kcfPath := filepath.Join(currentDir, "kubeconfig.yaml")
+	config, err := clientcmd.BuildConfigFromFlags("", kcfPath)
+	if err != nil {
+		return err
+	}
+	// Bypass TLS verification for vCluster - Required for Nodeport way to connect
+	config.Insecure = true
+	config.TLSClientConfig.CAFile = ""
+	config.TLSClientConfig.CAData = nil
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	v.Clientset = clientset
+	return nil
+}
+
+// CreatePVC creates a PVC in vcluster
+func (v *VCluster) CreatePVC(svcName, appNs string) (string, error) {
+	pvcName := v.Name + "-" + svcName + "-pvc"
+	createOpts := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pvcName,
+			Namespace: appNs,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			StorageClassName: &svcName,
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("5Gi"),
+				},
+			},
+		},
+	}
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: appNs,
+		},
+	}
+	// Creating Namespace in VCluster first before creating PVC
+	if _, err := v.Clientset.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{}); err != nil {
+		return "", err
+	}
+	_, err := v.Clientset.CoreV1().PersistentVolumeClaims(appNs).Create(context.TODO(), createOpts, metav1.CreateOptions{})
+	if err != nil {
+		return "", err
+	}
+	return pvcName, nil
+}
+
+// int32Ptr converts integre to pointer
+func int32Ptr(i int32) *int32 { return &i }
+
+// CreateFIODeployment creates a FIO Batch Job on given PVC
+func (v *VCluster) CreateFIODeployment(pvcName string, appNS string, fioOpts FIOOptions) error {
```

Comment:
Is there any reason for adding this method? Can't we use existing way to deploy the FIO pods or other applications specs in vCluster?

Reply:
Hi Smit, yes I considered this. The problem with trying to use existing specs is that they are quite tightly coupled to their own constructs. Now there are 2 things we need to consider : 
1. With vcluster few things need to be created at Host layer and few at vcluster layer
2. Torpedo doesn't support multiple k8s constructs. For every k8s construct, we need to do an InitInstance() and this initinstance will need to be modified when we are trying to do a vcluster init. It's much better, quicker and flexible if we have vcluster constructs separately rather than trying to intrusively modify current Torpedo.

Reply:
Can't we have vCluster scheduler driver which have it own node instances and when we deploying these app specs using vCluster scheduler instance that time it will deploy in vClutser.  

If required vCluster provisioning part can be moved to spawn.

Reply:
Discussed this - pasting summary here: 
I did consider this approach as I mentioned in the comment as well. The problem with this approach is it will cause intrusive changes in Torpedo. I don't believe that's the right thing to do in Torpedo when we have an equally great way out. The ROI in the other approach is only that we will be able to utilise existing app specs - which is not very flexible in vcluster context. We will get ROI when we do a lot of intrusive changes. Rest everything is anyways happening as is using Torpedo.
Regarding node instances of vcluster, initinstance will not work as is. We will have to make changes to make it work in vcluster context
Another thing is : Torpedo doesn't support more than one context at one time.